File: sdks/js/src/core/context.ts
import { AsyncLocalStorage } from 'node:async_hooks';
import { Event } from './event.ts';

const storage = new AsyncLocalStorage<Event>();

/** Store an event in the current async context. */
export function storeEvent(event: Event): void {
  storage.enterWith(event);
}

/** Get the event from the current async context, or undefined. */
export function getEvent(): Event | undefined {
  return storage.getStore();
}

/** Check if there's an event in the current context. */
export function hasEvent(): boolean {
  return storage.getStore() !== undefined;
}

/** Get the event ID from the current context. */
export function eventId(): string {
  return storage.getStore()?.eventId ?? '';
}

/** Run a function with a specific event in context. */
export function runWithEvent<T>(event: Event, fn: () => T): T {
  return storage.run(event, fn);
}

/** Get the request ID from the current context. */
export function requestIdFromContext(): string {
  return storage.getStore()?.requestId ?? '';
}

/** Get the trace ID from the current context. */
export function traceIdFromContext(): string {
  return storage.getStore()?.traceId ?? '';
}

File: sdks/go/src/core/context.go
package core

import "context"

// contextKey is a private type to prevent key collisions with other packages.
type contextKey struct{}

// eventKey is the singleton context key used to store *Event.
var eventKey = contextKey{}

// storeEvent stores ev in ctx and returns the derived context.
func storeEvent(ctx context.Context, ev *Event) context.Context {
	return context.WithValue(ctx, eventKey, ev)
}

// loadEvent returns the event from ctx, or nil when absent.
func loadEvent(ctx context.Context) *Event {
	ev, _ := ctx.Value(eventKey).(*Event)
	return ev
}

// FromContext retrieves the canonical Event from ctx.
// Returns (nil, false) if StartEvent has not been called.
func FromContext(ctx context.Context) (*Event, bool) {
	ev := loadEvent(ctx)
	return ev, ev != nil
}

// HasEvent reports if ctx has an active canonical event.
func HasEvent(ctx context.Context) bool {
	_, ok := FromContext(ctx)
	return ok
}

// EventID returns the event id from ctx when present.
func EventID(ctx context.Context) string {
	ev := loadEvent(ctx)
	if ev == nil {
		return ""
	}
	ev.MuLock()
	defer ev.MuUnlock()
	return ev.EventID
}

// RequestIDFromContext returns the request id from ctx when present.
func RequestIDFromContext(ctx context.Context) string {
	ev := loadEvent(ctx)
	if ev == nil {
		return ""
	}
	ev.MuLock()
	defer ev.MuUnlock()
	return ev.RequestID
}

// TraceIDFromContext returns the trace id from ctx when present.
func TraceIDFromContext(ctx context.Context) string {
	ev := loadEvent(ctx)
	if ev == nil {
		return ""
	}
	ev.MuLock()
	defer ev.MuUnlock()
	return ev.TraceID
}

// SpanIDFromContext returns the span id from ctx when present.
func SpanIDFromContext(ctx context.Context) string {
	ev := loadEvent(ctx)
	if ev == nil {
		return ""
	}
	ev.MuLock()
	defer ev.MuUnlock()
	return ev.SpanID
}
